hal/inc/sleep_hal.h

```diff
@@ -56,23 +61,24 @@ typedef enum hal_sleep_mode_t {
     HAL_SLEEP_MODE_MAX = 0x7F
 } hal_sleep_mode_t;
 
+// Bit mask enum value.
 typedef enum hal_wakeup_source_type_t {
-    HAL_WAKEUP_SOURCE_TYPE_UNKNOWN = 0,
-    HAL_WAKEUP_SOURCE_TYPE_GPIO = 1,
-    HAL_WAKEUP_SOURCE_TYPE_ADC = 2,
-    HAL_WAKEUP_SOURCE_TYPE_DAC = 3,
-    HAL_WAKEUP_SOURCE_TYPE_RTC = 4,
-    HAL_WAKEUP_SOURCE_TYPE_LPCOMP = 5,
-    HAL_WAKEUP_SOURCE_TYPE_UART = 6,
-    HAL_WAKEUP_SOURCE_TYPE_I2C = 7,
-    HAL_WAKEUP_SOURCE_TYPE_SPI = 8,
-    HAL_WAKEUP_SOURCE_TYPE_TIMER = 9,
-    HAL_WAKEUP_SOURCE_TYPE_CAN = 10,
-    HAL_WAKEUP_SOURCE_TYPE_USB = 11,
-    HAL_WAKEUP_SOURCE_TYPE_BLE = 12,
-    HAL_WAKEUP_SOURCE_TYPE_NFC = 13,
-    HAL_WAKEUP_SOURCE_TYPE_NETWORK = 14,
-    HAL_WAKEUP_SOURCE_TYPE_MAX = 0x7FFF
+    HAL_WAKEUP_SOURCE_TYPE_UNKNOWN = 0x00,
+    HAL_WAKEUP_SOURCE_TYPE_GPIO = 0x01,
+    HAL_WAKEUP_SOURCE_TYPE_ADC = 0x02,
+    HAL_WAKEUP_SOURCE_TYPE_DAC = 0x04,
+    HAL_WAKEUP_SOURCE_TYPE_RTC = 0x08,
+    HAL_WAKEUP_SOURCE_TYPE_LPCOMP = 0x10,
+    HAL_WAKEUP_SOURCE_TYPE_UART = 0x20,
+    HAL_WAKEUP_SOURCE_TYPE_I2C = 0x40,
+    HAL_WAKEUP_SOURCE_TYPE_SPI = 0x80,
+    HAL_WAKEUP_SOURCE_TYPE_TIMER = 0x100,
+    HAL_WAKEUP_SOURCE_TYPE_CAN = 0x200,
+    HAL_WAKEUP_SOURCE_TYPE_USB = 0x400,
+    HAL_WAKEUP_SOURCE_TYPE_BLE = 0x800,
+    HAL_WAKEUP_SOURCE_TYPE_NFC = 0x1000,
+    HAL_WAKEUP_SOURCE_TYPE_NETWORK = 0x2000,
+    HAL_WAKEUP_SOURCE_TYPE_MAX = 0x7FFFFFFF
```

Comment:
Any reason to make it 32-bit?

Reply:
See: https://github.com/particle-iot/device-os/blob/feature/sleep20/hal/src/nRF52840/sleep_hal.cpp#L98 and https://github.com/particle-iot/device-os/blob/feature/sleep20/hal/src/nRF52840/sleep_hal.cpp#L280 to identify the enabled wakeup sources.

Reply:
That can still be done using 16-bit enums, right..?

Reply:
Yes, but I'd like to reserve more bits for future use case.

Reply:
Oh, I missed that this is a bitmask now. I don't think that's the best solution. What exactly is the problem in handling normal non-bitmasked ids?

Reply:
Making them as bit mask values is to replace the `exit_conditions` that is used to identify the wakeup sources. I don't want to new a bit-mask enum type to introduce complexity. So I just make use of the wakeup source type and I think 32-bits is sufficient for iterating all possible wakeup source type.

Reply:
Ideally we should use the enum class bitwise, while that branch hasn't been merged. If making the wakeup source types as bit-masks makes sense to you, I'll update it in the enum class bitwise PR: https://github.com/particle-iot/device-os/pull/1978/files, the same as what I've done to `BleCharacteristicProperty`.

Reply:
Honestly, I still don't really understand why we need to use bitmasks here. The HAL API itself does not require in any way for the wakeup source types to be combined (ORed), so the only reason I see is we are doing this for convenience of the internal implementation. And even then, I'm not sure why we'd want that, as we always have access to the configuration object and can traverse the list of wakeup options anyway.

> So I just make use of the wakeup source type and I think 32-bits is sufficient for iterating all possible wakeup source type.

I wouldn't be so sure that 31 wakeup options will be enough to cover all platforms in some time.

Reply:
> as we always have access to the configuration object and can traverse the list of wakeup options anyway

That's exactly what I did before using the bit-mask. The reason why I switched to use bit-mask is that it can reduce the time to re-enter sleep again if the wakeup source is not as expected after exiting sleep mode. Searching `wakeupSourceTypes` in sleep_hal.cpp you'll find that it used very often. If we don't go with that way, we can of course define a bunch of functions like `isWakeupSourceGpioEnabled()` to iterate the wakeup source chain to check if certain wakeup source is enabled.

Reply:
Reduce by how much? :smile: Traversing a singly-linked list is just `O(n)`. You don't really need to traverse it multiple times. My main argument here still is that we are influencing the API just for the sake of simplifying the internal implementation while also limiting ourselves to just 31 wakeup options. Wasn't one of our goals to have a flexible and easily portable HAL API?

Reply:
> You don't really need to traverse it multiple times
So simply define several bool variables instead to store the traversed result?

Reply:
Or better yet, can we simply configure things while we traverse the list? So:
1. Traverse the list and configure necessary peripherals for wakeup
2. Go into sleep
3. Wake up, figure out what woke us up
4. Traverse the list once again and deconfigure the peripherals that were used as wakeup sources